File: suspense-app/src/services/ingest.service.ts
import type { DB } from "@/repositories/_db";
import { getAccessibleAccountCodes } from "@/lib/db";
import { accountRepo } from "@/repositories/account.repo";
import { bankFormatRepo } from "@/repositories/bankFormat.repo";
import { balanceRepo } from "@/repositories/balance.repo";
import { suspenseRepo } from "@/repositories/suspense.repo";
import { getParser } from "./parsers/registry";
import { applyMapping } from "./mapping";
import type { ImportResult, RowError } from "@/domain/ingest.types";

export interface IngestInput {
  fileName: string;
  content: Buffer;
  screenBalanceDate: string; // 畫面餘額日期（檢核檔內日期須一致）
  userId: number;
  secondImport: boolean; // ☑ 二次轉入
  operator: string;
}

function nextBusinessDay(iso: string): string {
  const d = new Date(iso + "T00:00:00");
  do { d.setDate(d.getDate() + 1); } while (d.getDay() === 0 || d.getDay() === 6);
  return d.toISOString().slice(0, 10);
}

export interface PreviewResult {
  fileName: string;
  accountCode: string | null;
  profileId: number | null;
  records: import("@/domain/ingest.types").NormalizedBalanceRecord[];
  errors: RowError[];
  error?: string;
}

/** 試轉預覽（dry-run）：路由→profile→解析→對應，只回結果不寫入。 */
export function previewFile(db: DB, fileName: string, content: Buffer): PreviewResult {
  const base: PreviewResult = { fileName, accountCode: null, profileId: null, records: [], errors: [] };
  const matched = accountRepo.findByImportFileName(db, fileName);
  if (matched.length === 0) return { ...base, error: "檔名比對不到帳號" };
  if (matched.length > 1) return { ...base, error: "檔名命中多個帳號" };
  const account = matched[0];
  const accountCurrency = (account.currency_type as string) === "TWD" ? "NTD" : "ZZZ";
  const profile = bankFormatRepo.resolve(db, account.bank_code as string, accountCurrency);
  if (!profile) return { ...base, accountCode: account.account_code as string, error: `查無格式設定（銀行 ${account.bank_code}）` };
  try {
    const rows = getParser(profile.engine)(content, profile);
    const mapped = applyMapping(rows, profile, { currency: accountCurrency === "NTD" ? "NTD" : undefined });
    return { fileName, accountCode: account.account_code as string, profileId: profile.id, records: mapped.records, errors: mapped.errors };
  } catch (e) {
    return { ...base, accountCode: account.account_code as string, profileId: profile.id, error: e instanceof Error ? e.message : "解析失敗" };
  }
}

/** 轉入單一檔案（一檔一帳號）。回傳該檔的轉檔結果；逐檔獨立，不丟例外讓整批中斷。 */
export function ingestFile(db: DB, input: IngestInput): ImportResult {
  const fail = (msg: string, extra?: Partial<ImportResult>): ImportResult => ({
    fileName: input.fileName, accountCode: null, profileId: null, balanceDate: input.screenBalanceDate,
    total: 0, success: 0, fail: 1, status: "FAILED", errors: [{ sourceRow: 0, message: msg }], ...extra,
  });

  // 1. 路由：檔名 → 單一帳號
  const matched = accountRepo.findByImportFileName(db, input.fileName);
  if (matched.length === 0) return fail("檔名比對不到帳號，請於帳號基本資料設定 import_file_name 或手動指定");
  if (matched.length > 1) return fail("檔名命中多個帳號，請手動指定");
  const account = matched[0];
  const accountCode = account.account_code as string;
  const bankCode = account.bank_code as string;
  const accountCurrency = (account.currency_type as string) === "TWD" ? "NTD" : "ZZZ";

  // 2. 權限：帳號須在操作者可維護範圍
  const accessible = getAccessibleAccountCodes(db, input.userId, input.screenBalanceDate);
  if (accessible !== null && !accessible.includes(accountCode)) {
    return fail("您沒有此帳號的維護權限", { accountCode });
  }

  // 3. 解析 profile（銀行＋幣別，ZZZ fallback）
  const profile = bankFormatRepo.resolve(db, bankCode, accountCurrency);
  if (!profile) return fail(`查無格式設定（銀行 ${bankCode}）`, { accountCode });

  // 4. Parse + Map
  let records, mapErrors: RowError[];
  try {
    const rows = getParser(profile.engine)(input.content, profile);
    const mapped = applyMapping(rows, profile, { currency: accountCurrency === "NTD" ? "NTD" : undefined });
    records = mapped.records;
    mapErrors = mapped.errors;
  } catch (e) {
    return fail(e instanceof Error ? e.message : "解析失敗", { accountCode, profileId: profile.id });
  }

  const errors: RowError[] = [...mapErrors];
  let success = 0;

  // 候選暫收日期（供「已立暫收則擋覆蓋」檢核）：當日(二次) + 次營業日(日常)
  const suspenseDates = [input.screenBalanceDate, nextBusinessDay(input.screenBalanceDate)];

  const writeAll = db.transaction(() => {
    for (const r of records) {
      // 檢核：檔內日期須與畫面一致
      if (r.balanceDate !== input.screenBalanceDate) {
        errors.push({ sourceRow: r.sourceRow, field: "餘額日期", message: `檔內日期 ${r.balanceDate} 與畫面餘額日期 ${input.screenBalanceDate} 不一致` });
        continue;
      }
      // 檢核：檔內帳號（若有）須與路由帳號一致
      if (r.accountCode && r.accountCode !== accountCode) {
        errors.push({ sourceRow: r.sourceRow, field: "帳號", message: `檔內帳號 ${r.accountCode} 與檔名對應帳號 ${accountCode} 不一致` });
        continue;
      }
      const currency = r.currency;
      const maxSeq = balanceRepo.getMaxSeq(db, r.balanceDate, accountCode, currency);

      if (input.secondImport) {
        // 二次轉入：保留前筆，次別 +1
        balanceRepo.insertFileImport(db, {
          balanceDate: r.balanceDate, accountCode, currency, balance: r.balance,
          importSeq: maxSeq + 1, fileName: input.fileName, createdBy: input.operator,
        });
        success++;
      } else if (maxSeq === 0) {
        // 首次轉入
        balanceRepo.insertFileImport(db, {
          balanceDate: r.balanceDate, accountCode, currency, balance: r.balance,
          importSeq: 1, fileName: input.fileName, createdBy: input.operator,
        });
        success++;
      } else {
        // 更正覆蓋：先檢核是否已立暫收
        if (suspenseRepo.existsByAccountCurrencyDates(db, accountCode, currency, suspenseDates)) {
          errors.push({ sourceRow: r.sourceRow, field: "立暫收", message: "該帳號餘額已被立暫收，須先取消立暫收才能重轉" });
          continue;
        }
        const latest = balanceRepo.getLatestFileImport(db, r.balanceDate, accountCode, currency);
        if (latest) {
          balanceRepo.overwriteById(db, latest.id as number, { balance: r.balance, fileName: input.fileName, updatedBy: input.operator });
        } else {
          // 既有為手動/前日（非 FILE_IMPORT）：不覆蓋，另開新次別
          balanceRepo.insertFileImport(db, {
            balanceDate: r.balanceDate, accountCode, currency, balance: r.balance,
            importSeq: maxSeq + 1, fileName: input.fileName, createdBy: input.operator,
          });
        }
        success++;
      }
    }
  });
  writeAll();

  const total = success + errors.length;
  const status: ImportResult["status"] = success === 0 ? "FAILED" : errors.length > 0 ? "PARTIAL" : "SUCCESS";

  return {
    fileName: input.fileName, accountCode, profileId: profile.id, balanceDate: input.screenBalanceDate,
    total, success, fail: errors.length, status, errors,
  };
}

File: CashManagement/Services/IngestService.cs
using System.Text.Json;
using CashManagement.Data;
using CashManagement.Domain;
using CashManagement.Services.Parsing;
using Microsoft.EntityFrameworkCore;

namespace CashManagement.Services;

public class IngestInput
{
    public string FileName { get; init; } = "";
    public byte[] Content { get; init; } = [];
    public string ScreenBalanceDate { get; init; } = ""; // 畫面餘額日期（檢核檔內日期須一致）
    public int UserId { get; init; }
    public bool SecondImport { get; init; }              // ☑ 二次轉入
    public string Operator { get; init; } = "User";
}

/// <summary>
/// FUN2.1.1 轉檔三段式：Parse → Map → Validate+Write。
/// 一檔一帳號（檔名路由）；逐檔獨立、不互相中斷。
/// </summary>
public class IngestService(IDbContextFactory<AppDbContext> factory)
{
    /// <summary>多檔上傳：檢核重複檔名 → 逐檔轉入 → 寫轉檔歷程。</summary>
    public List<ImportResult> IngestFiles(List<(string FileName, byte[] Content)> files, string balanceDate, int userId, bool secondImport, string operatorName)
    {
        if (files.Count == 0) throw new BusinessException("未選擇檔案");
        if (string.IsNullOrEmpty(balanceDate)) throw new BusinessException("餘額日期必填");

        // 同批多檔不得有重複檔名
        var dup = files.GroupBy(f => f.FileName).FirstOrDefault(g => g.Count() > 1);
        if (dup != null) throw new BusinessException($"上傳檔案有重複檔名：{dup.Key}");

        var batchId = $"IMP-{DateTimeOffset.Now.ToUnixTimeMilliseconds()}";
        var results = new List<ImportResult>();
        foreach (var (fileName, content) in files)
        {
            var result = IngestFile(new IngestInput
            {
                FileName = fileName, Content = content, ScreenBalanceDate = balanceDate,
                UserId = userId, SecondImport = secondImport, Operator = operatorName,
            });
            WriteLog(batchId, result, operatorName);
            results.Add(result);
        }
        return results;
    }

    /// <summary>試轉預覽（dry-run）：路由→profile→解析→對應，只回結果不寫入。</summary>
    public PreviewResult PreviewFile(string fileName, byte[] content)
    {
        using var db = factory.CreateDbContext();
        var result = new PreviewResult { FileName = fileName };

        var matched = db.BankAccounts.AsNoTracking().Where(a => a.IsActive && a.ImportFileName == fileName).ToList();
        if (matched.Count == 0) { result.Error = "檔名比對不到帳號"; return result; }
        if (matched.Count > 1) { result.Error = "檔名命中多個帳號"; return result; }

        var account = matched[0];
        result.AccountCode = account.AccountCode;
        var accountCurrency = account.CurrencyType == "TWD" ? "NTD" : "ZZZ";

        var profile = ResolveProfile(db, account.BankCode, accountCurrency);
        if (profile == null) { result.Error = $"查無格式設定（銀行 {account.BankCode}）"; return result; }
        result.ProfileId = profile.Id;

        try
        {
            var rows = ParserRegistry.GetParser(profile.Engine)(content, profile);
            var (records, errors) = MappingHelper.ApplyMapping(rows, profile, accountCurrency == "NTD" ? "NTD" : null);
            result.Records = records;
            result.Errors = errors;
        }
        catch (Exception e)
        {
            result.Error = e.Message;
        }
        return result;
    }

    /// <summary>轉入單一檔案。回傳該檔結果；不拋例外讓整批中斷。</summary>
    public ImportResult IngestFile(IngestInput input)
    {
        using var db = factory.CreateDbContext();

        ImportResult Fail(string msg, string? accountCode = null, int? profileId = null) => new()
        {
            FileName = input.FileName, AccountCode = accountCode, ProfileId = profileId,
            BalanceDate = input.ScreenBalanceDate, Total = 0, Success = 0, Fail = 1,
            Status = "FAILED", Errors = [new RowError(0, null, msg)],
        };

        // 1. 路由：檔名 → 單一帳號
        var matched = db.BankAccounts.AsNoTracking().Where(a => a.IsActive && a.ImportFileName == input.FileName).ToList();
        if (matched.Count == 0) return Fail("檔名比對不到帳號，請於帳號基本資料設定轉檔檔名或手動指定");
        if (matched.Count > 1) return Fail("檔名命中多個帳號，請手動指定");
        var account = matched[0];
        var accountCurrency = account.CurrencyType == "TWD" ? "NTD" : "ZZZ";

        // 2. 權限：帳號須在操作者可維護範圍
        var accessible = PermissionService.GetAccessibleAccountCodes(db, input.UserId, input.ScreenBalanceDate);
        if (accessible != null && !accessible.Contains(account.AccountCode))
            return Fail("您沒有此帳號的維護權限", account.AccountCode);

        // 3. 解析 profile（銀行＋幣別，ZZZ fallback）
        var profile = ResolveProfile(db, account.BankCode, accountCurrency);
        if (profile == null) return Fail($"查無格式設定（銀行 {account.BankCode}）", account.AccountCode);

        // 4. Parse + Map
        List<NormalizedBalanceRecord> records;
        List<RowError> errors;
        try
        {
            var rows = ParserRegistry.GetParser(profile.Engine)(input.Content, profile);
            (records, errors) = MappingHelper.ApplyMapping(rows, profile, accountCurrency == "NTD" ? "NTD" : null);
        }
        catch (Exception e)
        {
            return Fail(e.Message, account.AccountCode, profile.Id);
        }

        var success = 0;

        // 候選暫收日期（供「已立暫收則擋覆蓋」檢核）：當日(二次) + 次營業日(日常)
        var suspenseDates = new[] { input.ScreenBalanceDate, SuspenseService.NextBusinessDay(db, input.ScreenBalanceDate) };

        using var txn = db.Database.BeginTransaction();
        foreach (var r in records)
        {
            // 檢核：檔內日期須與畫面一致
            if (r.BalanceDate != input.ScreenBalanceDate)
            {
                errors.Add(new RowError(r.SourceRow, "餘額日期", $"檔內日期 {r.BalanceDate} 與畫面餘額日期 {input.ScreenBalanceDate} 不一致"));
                continue;
            }
            // 檢核：檔內帳號（若有）須與路由帳號一致
            if (r.AccountCode.Length > 0 && r.AccountCode != account.AccountCode)
            {
                errors.Add(new RowError(r.SourceRow, "帳號", $"檔內帳號 {r.AccountCode} 與檔名對應帳號 {account.AccountCode} 不一致"));
                continue;
            }

            var maxSeq = db.PassbookBalances
                .Where(p => p.BalanceDate == r.BalanceDate && p.AccountCode == account.AccountCode && p.Currency == r.Currency)
                .Select(p => (int?)p.ImportSeq).Max() ?? 0;

            if (input.SecondImport)
            {
                // 二次轉入：保留前筆，次別 +1
                db.PassbookBalances.Add(NewImport(r, account.AccountCode, maxSeq + 1, input));
                success++;
            }
            else if (maxSeq == 0)
            {
                // 首次轉入
                db.PassbookBalances.Add(NewImport(r, account.AccountCode, 1, input));
                success++;
            }
            else
            {
                // 更正覆蓋：先檢核是否已立暫收
                var hasSuspense = db.SuspenseTransactions.Any(t =>
                    t.AccountCode == account.AccountCode && t.Currency == r.Currency && suspenseDates.Contains(t.SuspenseDate));
                if (hasSuspense)
                {
                    errors.Add(new RowError(r.SourceRow, "立暫收", "該帳號餘額已被立暫收，須先取消立暫收才能重轉"));
                    continue;
                }
                var latest = db.PassbookBalances
                    .Where(p => p.BalanceDate == r.BalanceDate && p.AccountCode == account.AccountCode
                                && p.Currency == r.Currency && p.DataType == BalanceDataTypes.FileImport)
                    .OrderByDescending(p => p.ImportSeq).FirstOrDefault();
                if (latest != null)
                {
                    latest.Balance = r.Balance;
                    latest.FileName = input.FileName;
                    latest.DataType = BalanceDataTypes.FileImport;
                    latest.IsReviewed = false;
                    latest.ReviewedBy = null;
                    latest.ReviewedAt = null;
                    latest.UpdatedBy = input.Operator;
                    latest.UpdatedAt = DateTime.Now;
                }
                else
                {
                    // 既有為手動/前日（非 FILE_IMPORT）：不覆蓋，另開新次別
                    db.PassbookBalances.Add(NewImport(r, account.AccountCode, maxSeq + 1, input));
                }
                success++;
            }
            db.SaveChanges();
        }
        txn.Commit();

        var total = success + errors.Count;
        var status = success == 0 ? "FAILED" : errors.Count > 0 ? "PARTIAL" : "SUCCESS";

        return new ImportResult
        {
            FileName = input.FileName, AccountCode = account.AccountCode, ProfileId = profile.Id,
            BalanceDate = input.ScreenBalanceDate, Total = total, Success = success,
            Fail = errors.Count, Status = status, Errors = errors,
        };
    }

    /// <summary>依「銀行＋幣別」解析有效 profile：先精確幣別、再 ZZZ fallback；取最新版本。</summary>
    private static ParsedProfile? ResolveProfile(AppDbContext db, string bankCode, string currency)
    {
        BankFormatProfile? Pick(string cur) => db.BankFormatProfiles.AsNoTracking()
            .Where(p => p.BankCode == bankCode && p.Currency == cur && p.Status == "ACTIVE")
            .OrderByDescending(p => p.Version).FirstOrDefault();

        var raw = Pick(currency) ?? (currency != "ZZZ" ? Pick("ZZZ") : null);
        return raw == null ? null : ParsedProfile.From(raw);
    }

    private static PassbookBalance NewImport(NormalizedBalanceRecord r, string accountCode, int seq, IngestInput input) => new()
    {
        BalanceDate = r.BalanceDate, AccountCode = accountCode, Currency = r.Currency, Balance = r.Balance,
        DataType = BalanceDataTypes.FileImport, ImportSeq = seq, FileName = input.FileName,
        IsReviewed = false, CreatedBy = input.Operator, UpdatedBy = input.Operator,
    };

    private void WriteLog(string batchId, ImportResult r, string uploadedBy)
    {
        using var db = factory.CreateDbContext();
        db.ImportLogs.Add(new ImportLog
        {
            BatchId = batchId, FileName = r.FileName, AccountCode = r.AccountCode, ProfileId = r.ProfileId,
            BalanceDate = r.BalanceDate, TotalCount = r.Total, SuccessCount = r.Success, FailCount = r.Fail,
            Status = r.Status, Errors = r.Errors.Count > 0 ? JsonSerializer.Serialize(r.Errors) : null,
            UploadedBy = uploadedBy,
        });
        db.SaveChanges();
    }
}
